1.8 - changes and bug fixes

artemkosenko opened the pull request:
Hi, Alex.

Правка скрипта. Ниже выдержка из того, что в сам скрипт добавлено.
Касательно 12 версии. Я несколько раз перепроверял. У меня все время получалась длина 492 байта. 
#1.8:  Added CMS R18 Support.
# Fixed a bug with bits unpacking format for CMS R16.x and R17
# Fixed a bug with CMS R16.x ASAI_UUI field length (changed from 96 to 97 bytes)
# Fixed a bug with V12 (CMS R12 - R15) record length (changed from 493 to 492 bytes)

Commit:
1.8 - changes and bug fixes

Commit:
V170 ASAI_UUI length fix

upinget commented:
Hi, is it possible to modify the script to convert the ASAI_UUI item to hex format because the binary data contains non printable characters which makes the paring very difficult.

artemkosenko commented:
Hi, it is possible and actually a good idea. (i had the similar issue recently)

artemkosenko committed:
Added option to specify ASAI_UUI output format (ASCII or HEX)

artemkosenko commented:
Now it is possible convert ASAI_UUI to hex format by specifying -x command line argument.
Tested on CMS R17.

upinget commented:
Thanks

artemkosenko committed:
added R19 support.
R19 version number - 190
ECHI doc states: "The VERSION field will contain the value 180 for every 
file header." which is strange.

artemkosenko committed:
upd: version log

sunnikolay commented:
Использую версию CMS R17, в колонки ASSIST, AUDIO, CONFERENCE, MALICIOUS, OBSERVINGCALL, TRANSFERRED, AGT_RELEASED данные всегда устанавливаются в "1", в колонках AGENTSURPLUS, AGENTSKILLLEVEL, PREFSKILLLEVEL данные не соответсвуют действительности.
Мы используем два сервера CMS, они находятся в разных локациях, используют "High Availibility". На первом CMS данные отправляются на внешнюю базу (использую echi-decode.pl), второй сервер складывает данные во внутреннюю базу данных.
Я сравнил данные полученные из 1 CMS с данными полученными из 2 CMS, и, они не совпадают, прикладываю файл вызовов. 
1) Строки в файле без заливки - это данные полученные через echi-decode.pl
2) Строки с желтой заливкой - это данные полученные из локальной базы 2 CMS
3) Черные ячейки не проверял
4) Красные ячейки - это несоответсвие данных echi-decode.pl и данных из 2 CMS(локальная база данных CMS).
5) В этих данных разные CALLID, я не знаю почему так, но одинаковые UCID.
[4401_1200.xlsx](https://github.com/nohuhu/echi-decode/files/3772885/4401_1200.xlsx)

artemkosenko committed:
Update echi-decode.pl

fixed R19 version

nohuhu commented on echi-decode.pl:
```diff
@@ -298,48 +343,82 @@ END
   12 =>		# CMS R12 to R15
   {
-    length => 493,
+    length => 492,
```
Вот эта строка вызывает у меня сомнения... С системами CMS я не работал уже более 15 лет, но когда этот скрипт разрабатывался для моих клиентов, версии 13-15 были как раз рабочими и использовались. Я не припомню проблем с конвертацией данных, которые непременно появились бы при такой ошибке.

Можете поделиться основаниями для этого исправления?

artemkosenko replied:
Сам давно не трогал CMS.
По документации разобрал каждую версию.
Свёл в таблицу, которая считает позиции и смещения.
Сумма байт для CMS 12-15 получилась 492.
![image](https://github.com/user-attachments/assets/1971f528-4847-4050-b012-cadc811805fc)

nohuhu reviewed (commented):
Ух, вот это я проморгал капитально... PR открыт почти 10 лет назад. :( Приношу свои извинения.

@artemkosenko Подскажите пожалуйста, вы до сих пор используете этот скрипт? У меня уже более 15 лет нет доступа к системам CMS, и я не могу проверить, работает ли скрипт с новыми версиями CMS. Я готов принять ваше слово, что всё работает, но хотелось бы уточнить состояние на текущий момент.
